### PR TITLE
[Search Relevance]  Refine the license compliance error message to be specific for features

### PR DESCRIPTION
**Pull request by saikatsarkar056**
Error message for features included in the Enterprise Search plugin is generic. In this PR, we modify this error message to include specific name of feature. Following is an example of updated error message.

# Before
> Current license is non-compliant for search application and behavioral analytics. Current license is null. Search Applications and behavioral analytics require an active trial, platinum or enterprise license.

# After 
> Current license is non-compliant for behavioral analytics. Current license is null.  This feature requires an active trial, platinum or enterprise license.

### CODE REVIEW

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearchFeature.java
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearchFeature.java:10]: I understand your instinct to make this strongly typed, but I have a couple of thoughts. 

First, I don't think we want the names of each individual API capitalized. So instead of `Search Application` I might suggest `search application` - this is more consistent with how we return error messages elsewhere as well as our documentation. 

Second, I'm not opposed to making this an enum, but I think it should be a top level enum and not part of an otherwise empty class. (See `QueryRuleCriteriaType` for an example of what I mean). Also could we come up with a more descriptive name than `EnterpriseSearchFeature`? I think `EnterpriseSearch` is redundant since we're already in the ent-search module. I suppose you could think about these as Products. You could also make a case for having static final Strings that we pass in as well.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/LicenseUtils.java
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/LicenseUtils.java:36]: I wonder if we could reword this a bit, especially for casing concerns? 

Would we be able to do something like `Current license is non-compliant for search applications. Current license is {}. This feature requires an active trial, platinum or enterprise license.`?

I'd want @ioanatia 's input on that though.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/action/RestDeleteAnalyticsCollectionAction.java
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/action/RestDeleteAnalyticsCollectionAction.java:28]: I wish there was a decent way to validate that we're passing in the correct value here, and not passing in say "Search applications" for an analytics call.

**Review by kderusso — COMMENTED** [review]
Hi Saikat, nice work so far. 

A couple of comments and questions in the review, also one other high level question: 
Your example showed a licenseState of `null`. I'm a bit concerned if that is a case we can run into?

File: docs/changelog/98409.yaml
Comment (saikatsarkar056) [reply docs/changelog/98409.yaml:2]: updated

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearchFeature.java
Comment (saikatsarkar056) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearchFeature.java:10]: What about `public enum ProductType` instead of `enum Feature`?

**Comment by saikatsarkar056** [comment]
@kderusso Is there any guidance around which changes should go to changelog?

**Comment by saikatsarkar056** [comment]
@kderusso Is there any guidance around which changes should go to changelog?

**Comment by saikatsarkar056** [comment]
> Hi Saikat, nice work so far.
> 
> A couple of comments and questions in the review, also one other high level question: Your example showed a licenseState of `null`. I'm a bit concerned if that is a case we can run into?

Can please redirect me to the point where licenseState can be `null`?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/LicenseUtils.java
Comment (ioanatia) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/LicenseUtils.java:36]: this message was following the initial product team requirements to not just have a general message that the license is non compliant, but to state the current license level, the required license level and the feature being used.
so I think the wording is fine, the casing should be lower case as it was initially.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/action/RestDeleteAnalyticsCollectionAction.java
Comment (ioanatia) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/action/RestDeleteAnalyticsCollectionAction.java:28]: we do have the rest action tests to account for this - not sure if there's something more here that is needed.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearchFeature.java
Comment (ioanatia) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearchFeature.java:10]: could we have the enum be defined as part of the `LicenseUtils`?
this enum is only used for licensing checks, so it would make sense to be part of `LicenseUtils`, instead of defining a standalone class.

Comment (saikatsarkar056) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearchFeature.java:10]: > You could also make a case for having static final Strings that we pass in as well.
@kderusso Can you please elaborate on this with an example? Do you mean to have static final Strings within the enum?

Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/EnterpriseSearchFeature.java:10]: Strings would replace an enum, however I like @ioanatia 's idea to move this to `LicenseUtils` - as a standalone class it's overkill.

**Comment by kderusso** [comment]
> @kderusso Is there any guidance around which changes should go to changelog?

@saikatsarkar056 [Here is some guidance](https://github.com/elastic/ml-cpp/blob/10f059d7ab12cd5e683b433fa672e7d706fd78ea/CONTRIBUTING.md?plain=1#L103). 

> Can please redirect me to the point where licenseState can be null?

My question was based off the example you provided in your PR description, where it says `Current license is null`.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/LicenseUtils.java
Comment (saikatsarkar056) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/LicenseUtils.java:36]: @ioanatia Do you mean that we can go with the new proposed exception message proposed by @kderusso?

**Comment by saikatsarkar056** [comment]
> > @kderusso Is there any guidance around which changes should go to changelog?
> 
> @saikatsarkar056 [Here is some guidance](https://github.com/elastic/ml-cpp/blob/10f059d7ab12cd5e683b433fa672e7d706fd78ea/CONTRIBUTING.md?plain=1#L103).
> 
> > Can please redirect me to the point where licenseState can be null?
> 
> My question was based off the example you provided in your PR description, where it says `Current license is null`.

O I see. I generated the new message from unit test. My understanding is that the license should not be null in production env.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/action/RestDeleteAnalyticsCollectionAction.java
Comment (saikatsarkar056) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/analytics/action/RestDeleteAnalyticsCollectionAction.java]: I have added additional unit tests to cover different exception messages.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/LicenseUtils.java
Comment (kderusso) [inline x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/LicenseUtils.java:19]: I apologize for the confusion here, when I mentioned that I liked Ioana's suggestion I thought that you were keeping it as an enum. Moving the enum wholesale made it a bit better in my eyes since we weren't creating a new class for the enum. Strings are OK too, but we weren't saying you had to do both. I'll leave that up to you. 

With that said I really dislike the name `Feature` here however it's used. I thought we'd talked about other names like `Product`?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/LicenseUtils.java
Comment (kderusso) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/LicenseUtils.java:37]: Please make sure to get this answered by @ioanatia before merging.

**Comment by kderusso** [comment]
> O I see. I generated the new message from the unit test. My understanding is that the license should not be null in production env.

@saikatsarkar056 Just to make sure, you've run this locally and observed the behavior with a basic license?

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/LicenseUtils.java
Comment (saikatsarkar056) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/LicenseUtils.java]: No worries. I changed it to `enum` now.

File: x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/LicenseUtils.java
Comment (ioanatia) [reply x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/utils/LicenseUtils.java:48]: current message looks okay with me - still fits the initial requirements from product.
if the feature name is lowercased, this is good to go

**Review by ioanatia — APPROVED** [review]

**Comment by saikatsarkar056** [comment]
> > O I see. I generated the new message from the unit test. My understanding is that the license should not be null in production env.
> 
> @saikatsarkar056 Just to make sure, you've run this locally and observed the behavior with a basic license?

My bad. I missed this comment. I will try to get a screenshot of exception message from locally run elasticsearch with basic license.

**Pull request merged**
